File: backend/app/models/habit.py
from sqlalchemy import Column, Integer, String, Text, Date, Boolean, ForeignKey

from app.database import Base


class Habit(Base):
    __tablename__ = "habits"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(200), nullable=False)
    description = Column(Text, nullable=True)
    frequency = Column(String(20), default="daily")  # daily / weekly
    target_count = Column(Integer, default=1)
    current_streak = Column(Integer, default=0)
    longest_streak = Column(Integer, default=0)
    color_theme = Column(String(20), default="blue")  # blue / green / orange / red
    is_archived = Column(Boolean, default=False)
    user_id = Column(Integer, ForeignKey("users.id"))

    # Audit defect #49: why the habit was archived + (optionally) the vault
    # document that triggered an automatic archive, so the Archive page can
    # explain the state instead of showing a bare "Archived" chip.
    archived_reason = Column(String(300), nullable=True)
    archived_document_id = Column(Integer, ForeignKey("kb_documents.id"), nullable=True)

    # Gamified Habit Tracker fields (99-phase plan, Phases 1-3)
    habit_type = Column(String(20), default="good")  # good / bad
    xp_reward = Column(Integer, default=30)
    xp_penalty = Column(Integer, default=20)
    image_url = Column(String(500), nullable=True)
    days_caught = Column(Integer, default=0)

    # Fitness Hub (Phase 3): goal text + cached 7x7 heatmap JSON
    goal = Column(String(300), nullable=True)
    heatmap_data = Column(Text, nullable=True)


class HabitLog(Base):
    __tablename__ = "habit_logs"

    id = Column(Integer, primary_key=True, index=True)
    habit_id = Column(Integer, ForeignKey("habits.id"))
    date = Column(Date, nullable=False)
    completed = Column(Boolean, default=False)
    count = Column(Integer, default=1)

    # Gamified Habit Tracker fields (Phase 4): Good/Bad, Completed/Shit, ±XP
    type = Column(String(10), default="good")  # good / bad
    status = Column(String(20), default="Completed")  # Completed / Shit I did it
    xp_change = Column(Integer, default=0)
    # Phase 91: drag-and-drop sort within a calendar day column
    sort_order = Column(Integer, default=0)

File: frontend/src/services/api/types/habit.ts
// Habit + pomodoro types.
// Extracted from services/api.ts (F8 split).

import type { Reward } from './rpg'

export interface Habit {
  id: number; name: string; description: string | null;
  frequency: string; target_count: number;
  current_streak: number; longest_streak: number; user_id: number;
  color_theme: string; is_archived: boolean;
  // Gamified Habit Tracker (Phases 1-3)
  habit_type: string; xp_reward: number; xp_penalty: number;
  image_url: string | null; days_caught: number;
  // Fitness Hub (Phase 3)
  goal: string | null; heatmap_data: string | null;
  // Archive provenance (defect #49).
  archived_reason?: string | null;
  archived_document_id?: number | null;
}

export interface HabitLog {
  id: number; habit_id: number; date: string;
  completed: boolean; count: number;
  // Gamified Habit Tracker (Phase 4)
  type: string; status: string; xp_change: number;
  // Phase 91: drag-and-drop sort within a calendar day column
  sort_order: number;
}

export interface HabitTodayItem {
  id: number; name: string; habit_type: string;
  xp_reward: number; xp_penalty: number; image_url: string | null;
  current_streak: number; days_caught: number;
  log_today: boolean; xp_today: number; status: string | null;
}

export interface HabitCalendarLog {
  id: number; habit_id: number; habit_name: string;
  habit_type: string; status: string; xp_change: number;
  count: number; completed: boolean; sort_order: number;
}

export interface HabitCalendarDay {
  date: string;
  logs: HabitCalendarLog[];
}

export interface HabitStats {
  records_this_month: number;
  days_missed: number;
  days_in_month: number;
  is_new_record: boolean;
  streak_graph: { date: string; streak_length: number }[];
}

export interface HabitHeatmapDay {
  date: string;
  completed: boolean;
  count: number;
}

export interface HabitHeatmap {
  month: string;
  days: HabitHeatmapDay[];
}

export interface PomodoroSession {
  id: number; user_id: number; start_time: string;
  duration_minutes: number; completed: boolean; task_description: string | null;
  // Gamified Habit Tracker (Phase 6)
  mode: string;
}

// ----- Gamified Habit Tracker types (99-phase plan) -----
export interface HabitTrackerCharacter {
  id: number; name: string;
  class_name: string; level: number; xp: number;
  avatar_class: string | null; current_streak: number;
}

export interface HabitTrackerTodayStat {
  id: number; habit_id: number; habit_name: string;
  habit_type: string; status: string; xp_change: number;
}

export interface HabitTrackerStatusWindow {
  character: HabitTrackerCharacter | null;
  xp_to_next: number | null;
  today_habits: HabitTrackerTodayStat[];
}

export interface HabitTrackerLifeArea {
  id: number; name: string; goal: string | null;
  image_url: string | null; status: string;
  progress_percent: number; sort_order: number;
  total_xp_earned: number;
}

export interface HabitTrackerSummary {
  total_xp: number;
  level: number;
  current_streak: number;
  good_today: number;
  bad_today: number;
  xp_to_next: number;
  good_count: number;
  bad_count: number;
  life_areas: HabitTrackerLifeArea[];
  rewards_available: Reward[];
}
